python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1611 @@
+  abstract ParameterNode getParameter(ParameterPosition ppos);
+
+  /** Gets the underlying library callable, if any. */
+  LibraryCallable asLibraryCallable() { this = TLibraryCallable(result) }
+
+  /** Gets the location of this dataflow callable. */
+  abstract Location getLocation();
+}
+
+/** A callable function. */
+abstract class DataFlowFunction extends DataFlowCallable, TFunction {
+  Function func;
+
+  DataFlowFunction() {
+    this = TFunction(func) and
+    // TODO: Handle @property decorators
+    not hasPropertyDecorator(func)
+  }
+
+  override string toString() { result = func.toString() }
+
+  override string getQualifiedName() { result = func.getQualifiedName() }
+
+  override Function getScope() { result = func }
+
+  override Location getLocation() { result = func.getLocation() }
+
+  /** Gets the positional parameter offset, to take into account self/cls parameters. */
+  int positionalOffset() { result = 0 }
+
+  override ParameterNode getParameter(ParameterPosition ppos) {
+    exists(int index | ppos.isPositional(index) |
+      result.getParameter() = func.getArg(index + this.positionalOffset())
+    )
+    or
+    exists(string name | ppos.isKeyword(name) | result.getParameter() = func.getArgByName(name))
+    or
+    // `*args`
+    exists(int index |
+      (
+        ppos.isStarArgs(index) and
+        result.getParameter() = func.getVararg()
+        or
+        ppos.isSynthStarArgsElement(index) and
+        result = TSynthStarArgsElementParameterNode(this)
+        or
+        ppos.isSynthLateStarArgsParameterPosition(index) and
+        result = TSynthLateStarArgsParameterNode(this)
+      )
+    |
+      // a `*args` parameter comes after the last positional parameter. We need to take
+      // self parameter into account, so for
+      // `def func(foo, bar, *args)` it should be index 2 (pos-param-count == 2)
+      // `class A: def func(self, foo, bar, *args)` it should be index 2 (pos-param-count - 1 == 3 - 1)
+      index = func.getPositionalParameterCount() - this.positionalOffset()
+      or
+      // no positional argument
+      not exists(func.getArg(_)) and index = 0
+    )
+    or
+    // `**kwargs`
+    // since dataflow library has restriction that we can only have ONE result per
+    // parameter position, if there is both a synthetic **kwargs and a real **kwargs
+    // parameter, we only give the result for the synthetic, and add local flow from the
+    // synthetic to the real. It might seem more natural to do it in the other
+    // direction, but since we have a clearStep on the real **kwargs parameter, we that
+    // content-clearing would also affect the synthetic parameter, which we don't want.
+    (
+      not exists(func.getArgByName(_)) and
+      ppos.isDictSplat() and
+      result.getParameter() = func.getKwarg()
+      or
+      exists(func.getArgByName(_)) and
+      ppos.isDictSplat() and
+      result = TSynthDictSplatParameterNode(this)
+    )
```

Comment:
This might be a matter of taste, but I think this formulation makes it clear that we choose only one when the position is for dict splat.
```suggestion
    ppos.isDictSplat() and
    if exists(func.getArgByName(_)) then
      result = TSynthDictSplatParameterNode(this)
    else
      result.getParameter() = func.getKwarg()
```

Reply:
I formatted this suggestion, so it was easier to accept straight away.

```suggestion
    ppos.isDictSplat() and
    if exists(func.getArgByName(_))
    then result = TSynthDictSplatParameterNode(this)
    else result.getParameter() = func.getKwarg()
```